composer/trainer/trainer.py

```diff
@@ -108,6 +108,112 @@
 __all__ = ["Trainer"]
 
 
+class MissingArgumentException(ValueError):
+
+    def __init__(self, arg_name: str) -> None:
+        super().__init__((f"{arg_name} is a required argument and must be specified when constructing the "
+                          f"{Trainer.__name__} or when calling {Trainer.__name__}.{Trainer.fit.__name__}(). "
+                          f"To fix, please specify `arg_name` via {Trainer.__name__}({arg_name}=...) or "
+                          f"{Trainer.__name__}.{Trainer.fit.__name__}({arg_name}=...)."))
+
+
+def _scale_max_duration_by_ssr(scale_schedule_ratio: float,
+                               orig_max_duration: Optional[Time[int]]) -> Optional[Time[int]]:
+    if orig_max_duration is None:
+        return None
+    max_duration = cast(Time[int], orig_max_duration * scale_schedule_ratio)
+    log.info(f'max_duration changed from {orig_max_duration} to {max_duration}')
+    if max_duration.value == 0:
+        raise ValueError('Scale schedule has reduced the max_duration to 0. Set a higher ratio or use more epochs.')
+    return max_duration
+
+
```

Comment:
Thoughts on moving all of these `unpack` functions to a separate helper file and importing them here for readability purposes? This `trainer.py` file is almost 2k lines long - would be good to shorten it somehow.

Reply:
A good number of those lines are docstrings :)
But yes, I can move it to a private helper file.

Reply:
After further thought, I'd prefer to leave it in `trainer.py`. While it would help with limiting git diffs if they were in a separate file, I do think it's helpful to keep them here, as they do belong as part of the trainer (perhaps they should be private, static methods?)

I know the trainer is long, but not _unreasonably_ long. Both the PTL and HF training loops are on a similar order of magnitude.

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py
https://github.com/huggingface/transformers/blob/v4.18.0/src/transformers/trainer.py

FastAI is way shorter, but they "outsource" a good amount of logic to callbacks.

Reply:
Agreed in keep this together in one file.